lib/web_ui/lib/src/engine/canvaskit/text.dart

```diff
@@ -712,8 +755,15 @@ class CkParagraphBuilder implements ui.ParagraphBuilder {
     _styleStack.add(skStyle);
     _commands.add(_ParagraphCommand.pushStyle(ckStyle));
     if (skStyle.foreground != null || skStyle.background != null) {
-      final SkPaint foreground = skStyle.foreground?.skiaObject ?? _defaultTextStylePaint;
-      final SkPaint background = skStyle.background?.skiaObject ?? _defaultTextStylePaint;
+      SkPaint? foreground = skStyle.foreground?.skiaObject;
+      if (foreground == null) {
+        _defaultTextForeground.setColorInt(
+          skStyle.color?.value ?? 0xFF000000,
+        );
```

Comment:
I have a bad feeling about this mutation of a static default style. Doesn't it leak the color to other paragraphs?

Reply:
We haven't found a better way to do this yet. JavaScript doesn't have value types so any structs we pass to WASM end up extra objects allocated in the heap creating GC pressure. So in a few cases where we know it's safe, we keep a shared statically allocated WASM object and use it to pass data to CanvasKit without allocating extra objects. There are a few more examples in `canvaskit_api.dart` where we do this, such as:

https://github.com/flutter/engine/blob/5cf3eaed113e8d32f6e61396eb252d76ad72c5b5/lib/web_ui/lib/src/engine/canvaskit/canvaskit_api.dart#L1024

I tested this specifically by invoking it twice in a row and I didn't see it leaking to another text style. But I agree, that this can indeed be dangerous. You have to know that internally Skia is taking the object by value rather than by pointer. We should avoid this pattern unless there's a strong reason to use it.

Reply:
One case I can think of is:

1. First paragraph:
  * `foreground == null`
  * `skStyle.color.value == GREEN`
2. Second paragraph:
  * `foreground == null`
  * `skStyle.color.value == null`

**_Expected:_** The first paragraph is green, the second paragraph has the default color (black?).

**_Actual:_** The green color leaked to the second paragraph.


That said, I agree with the GC pressure issue. Maybe add a quick comment here explaining the possible leak and the reason it was done this way?

Reply:
Ok, I'm convinced that this is risky, so I added a test to make sure we don't leak styles: https://github.com/flutter/goldens/blob/76847ac130d9dc2ec4188694daff8fa16735a7d8/engine/web/canvaskit_text_styles_do_not_leak.png